### PR TITLE
16) added new models and connected with it services/Dao

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/ShoppingCartDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(shoppingCart);
+            transaction.commit();
+            return shoppingCart;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataBindingException("Can't add this shop cart to BD "
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/dao/impl/ShoppingCartDaoImpl.java]:
```suggestion
            throw new DataBindingException("Can't add to DB shop cart  "
```

File: src/main/java/com/dev/cinema/model/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.MapsId;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "shopping_cart")
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/model/ShoppingCart.java]:
```suggestion
@Table(name = "shopping_carts")
```
same with others

File: src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.dev.cinema.model.ShoppingCart;
+import com.dev.cinema.model.Ticket;
+import com.dev.cinema.model.User;
+import com.dev.cinema.service.ShoppingCartService;
+
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    @Inject private MovieSessionDao movieSessionDao;
+    @Inject private ShoppingCartDao shoppingCartDao;
+    @Inject private TicketDao ticketDao;
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    @Inject 
    private MovieSessionDao movieSessionDao;
    @Inject 
    private ShoppingCartDao shoppingCartDao;
    @Inject 
    private TicketDao ticketDao;
```

File: src/main/java/com/dev/cinema/Main.java [hunk 4]
```diff
@@ -44,11 +46,37 @@ public static void main(String[] args) {
...
+
+        ShoppingCartService shoppingCartService =
+                (ShoppingCartService) injector
+                        .getInstance(ShoppingCartService.class);
+        try {
+            shoppingCartService.getByUser(authenticationService
+                    .login("SomeEmail@mail.com",
+                            "SomePassword"));
+        } catch (AuthenticationException e) {
+            e.printStackTrace();
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/Main.java]: I would not recommend to try-catch all of these. Save the result of `login()` to a variable and use it in `shoppingCarService`. But if you still want to keep it as it is - `e.printStackTrace()` is a bad practice. You can show the AuthenticationException message

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/Main.java [hunk 5]
```diff
@@ -56,27 +56,16 @@ public static void main(String[] args) {
...
-            e.printStackTrace();
-        }
-        try {
-            shoppingCartService.addSession(movieSession, authenticationService
-                    .login("SomeEmail@mail.com",
-                            "SomePassword"));
+            userForTest = authenticationService
+                    .login("SomeEmail@mail.com", "SomePassword");
         } catch (AuthenticationException e) {
             e.printStackTrace();
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/Main.java:64]: still, `e.printStackTrace()` - bad practice

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/com/dev/cinema/dao/impl/ShoppingCartDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import com.dev.cinema.lib.Dao;
+import com.dev.cinema.model.ShoppingCart;
+import com.dev.cinema.model.User;
+import com.dev.cinema.util.HibernateUtil;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+@Dao
+public class ShoppingCartDaoImpl implements ShoppingCartDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/impl/ShoppingCartDaoImpl.java]: redundant empty line

File: src/main/java/com/dev/cinema/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import javax.persistence.MapsId;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/ShoppingCart.java]: is it needed?

File: src/main/java/com/dev/cinema/model/Ticket.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "ticket")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/Ticket.java]: let's use either singular or plural for all table names

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 9]
```diff
@@ -4,22 +4,23 @@
...
 import com.dev.cinema.service.UserService;
 import com.dev.cinema.util.HashUtil;
 import java.util.Optional;
 
 @Service
 public class AuthenticationServiceImpl implements AuthenticationService {
-    @Inject
-    private UserService userService;
+    @Inject private UserService userService;
+    @Inject private ShoppingCartService shoppingCartService;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
    @Inject 
    private UserService userService;
    @Inject 
    private ShoppingCartService shoppingCartService;
```

File: src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import com.dev.cinema.model.MovieSession;
+import com.dev.cinema.model.ShoppingCart;
+import com.dev.cinema.model.Ticket;
+import com.dev.cinema.model.User;
+import com.dev.cinema.service.ShoppingCartService;
+
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    @Inject
+    private MovieSessionDao movieSessionDao;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java]: do you use this?

File: src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    private ShoppingCartDao shoppingCartDao;
+    @Inject
+    private TicketDao ticketDao;
+
+    @Override
+    public void addSession(MovieSession movieSession, User user) {
+        Ticket ticket = new Ticket();
+        ticket.setUser(user);
+        ticket.setMovieSession(movieSession);
+        ticketDao.add(ticket);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java]: if there is no shopping cart for the user - we still save ticket to the db, is it ok?

File: src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    public void registerNewShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartDao.add(shoppingCart);
+    }
+
+    @Override
+    public void clear(ShoppingCart shoppingCart) {
+        User user = shoppingCart.getUser();
+        ShoppingCart shoppingCartByUser = getByUser(user);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/impl/ShoppingCartServiceImpl.java]: do you need this?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**